theano/sandbox/cpu_corr/blas.py

```diff
@@ -0,0 +1,499 @@
+#from theano import tensor
+from theano.compat.six import StringIO
+from theano import gof
+from theano.tensor import as_tensor_variable
+from theano.tensor.blas_headers import (blas_header_text,
+                                        blas_header_version)
+from theano.tensor.blas import ldflags
+
+
+class BaseCpuCorrMM(gof.Op):
+    """Base class for `CpuCorrMM`, `CpuCorrMM_gradWeights` and
+    `CpuCorrMM_gradInputs`. Cannot be used directly.
+
+    :param border_mode: one of 'valid', 'full', 'half'; additionally, the
+        padding size could be directly specified by an integer or a pair of
+        integers
+    :param subsample: perform subsampling of the output (default: (1, 1))
+    :param pad: *deprecated*, now you should always use border_mode
+
+    """
+    check_broadcast = False
+    __props__ = ('border_mode', 'subsample')
+
+    def __init__(self, border_mode="valid", subsample=(1, 1), pad=(0, 0)):
+        if pad != (0, 0):
+            _logger.warning(
+                'do not use pad for BaseCpuCorrMM; please set padding in '
+                'border_mode parameter, see the docstring for more details')
+            if border_mode != "valid":
+                raise ValueError("border_mode must be 'valid'")
+            border_mode = pad
+        if isinstance(border_mode, int):
+            border_mode = (border_mode, border_mode)
+        if isinstance(border_mode, tuple):
+            pad_h, pad_w = map(int, border_mode)
+            border_mode = (pad_h, pad_w)
+        if not ((isinstance(border_mode, tuple) and min(border_mode) >= 0) or
+                border_mode in ('valid', 'full', 'half')):
+            raise ValueError(
+                'invalid border_mode {}, which must be either '
+                '"valid", "full", "half", an integer or a pair of'
+                ' integers'.format(border_mode))
+        self.border_mode = border_mode
+        if len(subsample) != 2:
+            raise ValueError("subsample must have two elements")
+        self.subsample = subsample
+
+    @property
+    def pad(self):
+        if self.border_mode != 'valid':
+            return self.border_mode
+        return (0, 0)
+
+    def __str__(self):
+        return '%s{%s, %s}' % (
+            self.__class__.__name__,
+            self.border_mode,
+            str(self.subsample))
+
+    def flops(self, inp, outp):
+        """ Useful with the hack in profilemode to print the MFlops"""
+        # if the output shape is correct, then this gives the correct
+        # flops for any direction, sampling, padding, and border mode
+        inputs, filters = inp
+        outputs, = outp
+        assert inputs[1] == filters[1]
+        # nb mul and add by output pixel
+        flops = filters[2] * filters[3] * 2
+        # nb flops by output image
+        flops *= outputs[2] * outputs[3]
+        # nb patch multiplied
+        flops *= inputs[1] * filters[0] * inputs[0]
+        return flops
+
+    def c_support_code(self):
```

Comment:
I didn't know the COp interface, I just continued from what @stencilman had started in #2002. Not sure if there would be an advantage in refactoring it to be based on COp, I think it's fine to just use about the same Python code as for GpuCorrMM.

What I was referring to was that https://github.com/Theano/Theano/blob/master/theano/sandbox/cuda/corr_gemm.cu is based on https://github.com/BVLC/caffe/blob/master/src/caffe/layers/conv_layer.cu, and when porting this to CPU you might want to grab some parts from https://github.com/BVLC/caffe/blob/master/src/caffe/layers/conv_layer.cpp instead of translating the kernels yourself (acknowledging the source and original license as I did).